File: examples/customer-address-jpa/src/test/java/io/zenwave360/example/infrastructure/jpa/inmemory/CustomerRepositoryInMemory.java
package io.zenwave360.example.infrastructure.jpa.inmemory;

import static org.apache.commons.lang3.ObjectUtils.firstNonNull;

import io.zenwave360.example.core.domain.Customer;
import io.zenwave360.example.core.outbound.jpa.CustomerRepository;

public class CustomerRepositoryInMemory extends InMemoryJpaRepository<Customer> implements CustomerRepository {

    private long nextId = 0;
    private final PrimaryKeyGenerator<Long> primaryKeyGenerator = () -> nextId++;

    public Customer save(Customer entity) {
        entity = super.save(entity);
        entity.getPaymentMethods().forEach(paymentMethod -> {
            paymentMethod.setId(firstNonNull(paymentMethod.getId(), primaryKeyGenerator.next()));
        });
        return entity;
    }
}

File: examples/kustomer-address-jpa/src/test/kotlin/io/zenwave360/examples/kotlin/repository/jpa/inmemory/CustomerRepositoryInMemory.kt
package io.zenwave360.examples.kotlin.repository.jpa.inmemory

import io.zenwave360.examples.kotlin.domain.*
import io.zenwave360.examples.kotlin.repository.jpa.CustomerRepository
import org.apache.commons.lang3.ObjectUtils
import java.math.*
import java.time.*
import java.util.*

class CustomerRepositoryInMemory : InMemoryJpaRepository<Customer, Long>(), CustomerRepository {
    private var nextId: Long = 0
    private val primaryKeyGenerator: PrimaryKeyGenerator<Long?> = PrimaryKeyGenerator { nextId++ }

    override fun <S : Customer> save(entity: S): S {
        super.save(entity)
        entity.paymentMethods.forEach({ paymentMethod ->
            paymentMethod.id = ObjectUtils.firstNonNull(
                paymentMethod.id,
                primaryKeyGenerator.next()
            )
        })
        return entity
    }
}
